sia/synchronize.go

```diff
@@ -85,20 +92,20 @@ func (c *Core) CatchUp(peer network.Address) {
 
 	// unlock state during network I/O
 	err := peer.RPC("SendBlocks", blockArray, &newBlocks)
-	if err != nil {
+	if err != nil && err != moreBlocksErr {
 		// log error
+		// TODO: try a different peer?
 		return
 	}
 
-	prevHeight := c.Height()
 	for _, block := range newBlocks {
-		c.processBlock(block) // processBlock is a blocking function.
+		c.AcceptBlock(block)
 	}
 
 	// TODO: There is probably a better approach than to call CatchUp
 	// recursively. Furthermore, if there is a reorg that's greater than 100
 	// blocks, CatchUp is going to fail outright.
-	if prevHeight != c.Height() {
-		go c.CatchUp(c.RandomPeer())
+	if err == moreBlocksErr {
+		go c.CatchUp(peer)
```

Comment:
I'm reusing the old peer here, but that might not be best practice. On one hand, if a peer tells you it has more blocks, you probably want to ask that specific peer again. On the other hand, this allows a malicious host to trap you in an infinite CatchUp loop by always returning `moreBlocksErr`.